pkg/api/handlers/compat/images_build.go

```diff
@@ -920,6 +946,122 @@ func BuildImage(w http.ResponseWriter, r *http.Request) {
+	}
+
+	additionalContexts = make(map[string]*buildahDefine.AdditionalBuildContext)
+	for {
+		part, err := reader.NextPart()
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			return "", nil, fmt.Errorf("failed to read part: %w", err)
+		}
+
+		fieldName := part.FormName()
+		fileName := part.FileName()
+
+		logrus.Debugf("Processing part: field=%q, file=%q", fieldName, fileName)
+
+		switch {
+		case fieldName == "context" && fileName != "":
+			contextPath := filepath.Join(anchorDir, "context")
+			if err := os.MkdirAll(contextPath, 0755); err != nil {
+				part.Close()
+				return "", nil, fmt.Errorf("creating context directory: %w", err)
+			}
+
+			if err := archive.Untar(part, contextPath, nil); err != nil {
+				part.Close()
+				return "", nil, fmt.Errorf("extracting main context: %w", err)
+			}
+			contextDir = contextPath
+			logrus.Debugf("Extracted main context to %q", contextDir)
+			part.Close()
+
+		case strings.HasPrefix(fieldName, "buildcontext-url-"):
+			contextName := strings.TrimPrefix(fieldName, "buildcontext-url-")
+			urlBytes, err := io.ReadAll(part)
+			part.Close()
+			if err != nil {
+				return "", nil, fmt.Errorf("reading URL value: %w", err)
+			}
+			urlValue := string(urlBytes)
+
+			logrus.Debugf("Found URL context %q: %s", contextName, urlValue)
+
+			tempDir, subDir, err := buildahDefine.TempDirForURL(anchorDir, "buildah", urlValue)
+			if err != nil {
+				return "", nil, fmt.Errorf("downloading URL context %q: %w", contextName, err)
+			}
+
+			contextPath := filepath.Join(tempDir, subDir)
+			additionalContexts[contextName] = &buildahDefine.AdditionalBuildContext{
+				IsURL:           true,
+				IsImage:         false,
+				Value:           contextPath,
+				DownloadedCache: contextPath,
+			}
+			logrus.Debugf("Downloaded URL context %q to %q", contextName, contextPath)
+
+		case strings.HasPrefix(fieldName, "buildcontext-image-"):
+			contextName := strings.TrimPrefix(fieldName, "buildcontext-image-")
+			imageBytes, err := io.ReadAll(part)
+			part.Close()
+			if err != nil {
+				return "", nil, fmt.Errorf("reading image reference: %w", err)
+			}
+			imageRef := string(imageBytes)
+
+			logrus.Debugf("Found image context %q: %s", contextName, imageRef)
+
+			additionalContexts[contextName] = &buildahDefine.AdditionalBuildContext{
+				IsImage: true,
+				IsURL:   false,
+				Value:   imageRef,
+			}
+			logrus.Debugf("Added image context %q with reference %q", contextName, imageRef)
+
+		case strings.HasPrefix(fieldName, "buildcontext-local-") && fileName != "":
+			contextName := strings.TrimPrefix(fieldName, "buildcontext-local-")
+			additionalAnchor := filepath.Join(anchorDir, "additional", contextName)
+
+			if err := os.MkdirAll(additionalAnchor, 0700); err != nil {
+				part.Close()
+				return "", nil, fmt.Errorf("creating additional context directory %q: %w", contextName, err)
+			}
+
+			if err := archive.Untar(part, additionalAnchor, nil); err != nil {
```

Comment:
The archive content's coming from a client.  Why isn't it extracted using `chrootarchive.Untar()`, or `extractTarFile()`?